gobblin-core/src/main/java/gobblin/source/extractor/filebased/FileBasedExtractor.java

```diff
@@ -118,7 +118,8 @@ public D readRecordImpl(@Deprecated D reuse) throws DataRecordException, IOExcep
       readRecordStart = true;
     }
 
-    while ((currentFileItr == null || !currentFileItr.hasNext()) && !filesToPull.isEmpty()) {
+    boolean hasNext = currentFileItr.hasNext();
+    while ((currentFileItr == null || !hasNext) && !filesToPull.isEmpty()) {
```

Comment:
That's not correct because you have taken currentFileItr.hasNext() out of the while loop.